sidecar/register/registration.go

```diff
@@ -82,12 +83,14 @@ func (agent *RegistrationAgent) Stop() {
 
 func (agent *RegistrationAgent) register() {
 	for {
-		registeredInstance, err := agent.config.Client.Register(agent.config.ServiceInstance)
-		if err == nil {
-			go agent.renew(registeredInstance)
-			return
+		if agent.checkAppHealth(DefaultReregistrationDelay) {
```

Comment:
I was also considering that design. My concern was that health checks could be as stale the health check interval, somewhere between [0, 20s) in the case where the health check interval is the same as the default heartbeat interval. However, as you point out, it would decouple the health check and heartbeat intervals, which is probably of greater value than ensuring the health check is as fresh as possible.